content/tokio/topics/testing.md

```diff
@@ -0,0 +1,188 @@
+---
+title: "Unit Testing"
+---
+
+The purpose of this page is to give advice on how to write useful unit tests in
+asynchronous applications.
+
+## Pausing and resuming time in tests
+
+Sometimes, asynchronous code explicitly waits by calling [`tokio::time::sleep`]
+or waiting on a [`tokio::time::Interval::tick`]. Testing behaviour based on
+time (for example, an exponential backoff) can get cumbersome when the unit
+test starts running very slowly. However, internally, the time-related
+functionality of tokio supports pausing and resuming time. Pausing time has the
+effect that any time-related future may become ready early. The condition for
+the time-related future resolving early is that there are no more other futures
+which may become ready. This essentially fast-forwards time when the only
+future being awaited is time-related:
+
+```rust
+#[tokio::test]
+async fn paused_time() {
+    tokio::time::pause();
+    let start = std::time::Instant::now();
+    tokio::time::sleep(Duration::from_millis(500)).await;
+    println!("{:?}ms", start.elapsed().as_millis());
+}
+```
+
+This code prints `0ms` on a reasonable machine.
+
+For unit tests, it is often useful to run with paused time throughout. This can
+be achieved simply by setting the macro argument `start_paused` to `true`:
+
+```rust
+#[tokio::test(start_paused = true)]
+async fn paused_time() {
+    let start = std::time::Instant::now();
+    tokio::time::sleep(Duration::from_millis(500)).await;
+    println!("{:?}ms", start.elapsed().as_millis());
+}
+```
+
+See [tokio::test "Configure the runtime to start with time paused"](https://docs.rs/tokio/latest/tokio/attr.test.html#configure-the-runtime-to-start-with-time-paused) for more details.
+
+Of course, the temporal order of future resolution is maintained, even when
+using different time-related futures:
+
+```rust
+#[tokio::test(start_paused = true)]
+async fn interval_with_paused_time() {
+    let mut interval = interval(Duration::from_millis(300));
+    let _ = timeout(Duration::from_secs(1), async move {
+        loop {
+            interval.tick().await;
+            println!("Tick!");
+        }
+    })
+    .await;
+}
+```
+
+This code immediately prints `"Tick!"` exactly 4 times.
+
+[`tokio::time::Interval::tick`]: https://docs.rs/tokio/1/tokio/time/struct.Interval.html#method.tick
+[`tokio::time::sleep`]: https://docs.rs/tokio/1/tokio/time/fn.sleep.html
+
+## Mocking using [`AsyncRead`] and [`AsyncWrite`]
+
+The generic traits for reading and writing asynchronously ([`AsyncRead`] and
+[`AsyncWrite`]) are implemented by, for example, sockets. They can be used for
+mocking I/O performed by a socket.
+
+Consider, for setup, this simple TCP server loop:
+
+```rust
+# if true { return }
```

Comment:
... this is outside the main function. How can this work? There must be something wrong with our CI setup.

Reply:
... it's been a long day?! fixed